File: market-data/crypto-news/examples/javascript/blockchain.js
/**
 * Free Crypto News API - Blockchain & On-Chain Examples
 * https://github.com/nirholas/free-crypto-news
 * 
 * On-chain data, DeFi TVL, NFTs, security, and more.
 */

const BASE_URL = 'https://cryptocurrency.cv';

// =============================================================================
// ON-CHAIN DATA
// =============================================================================

/**
 * Get on-chain data for a blockchain
 * @param {string} chain - Blockchain name (ethereum, bitcoin, etc.)
 * @returns {Promise<Object>} On-chain metrics
 */
async function getOnChainData(chain) {
  const response = await fetch(`${BASE_URL}/api/blockchain/onchain?chain=${chain}`);
  return response.json();
}

/**
 * Get current gas prices
 * @param {string} chain - Blockchain (default: ethereum)
 * @returns {Promise<Object>} Gas price data
 */
async function getGasPrices(chain = 'ethereum') {
  const response = await fetch(`${BASE_URL}/api/blockchain/gas?chain=${chain}`);
  return response.json();
}

/**
 * Get token holder distribution
 * @param {string} token - Token address or symbol
 * @param {string} chain - Blockchain
 * @returns {Promise<Object>} Holder distribution
 */
async function getTokenHolders(token, chain = 'ethereum') {
  const response = await fetch(
    `${BASE_URL}/api/blockchain/holders?token=${token}&chain=${chain}`
  );
  return response.json();
}

/**
 * Get whale movements (large transactions)
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Whale alerts
 */
async function getWhaleMovements(options = {}) {
  const params = new URLSearchParams({
    chain: options.chain || 'all',
    minValue: options.minValue || 1000000,
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/whales?${params}`);
  return response.json();
}

// =============================================================================
// DEFI DATA
// =============================================================================

/**
 * Get DeFi Total Value Locked (TVL)
 * @param {Object} options - Query options
 * @returns {Promise<Object>} TVL data
 */
async function getDefiTVL(options = {}) {
  const params = new URLSearchParams({
    protocol: options.protocol || '',
    chain: options.chain || 'all',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/defi/tvl?${params}`);
  return response.json();
}

/**
 * Get yield farming opportunities
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Yield data
 */
async function getYields(options = {}) {
  const params = new URLSearchParams({
    chain: options.chain || 'all',
    minApy: options.minApy || 0,
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/defi/yields?${params}`);
  return response.json();
}

/**
 * Get staking information
 * @param {string} asset - Asset symbol
 * @returns {Promise<Object>} Staking data
 */
async function getStaking(asset) {
  const response = await fetch(`${BASE_URL}/api/blockchain/staking?asset=${asset}`);
  return response.json();
}

// =============================================================================
// TOKENS
// =============================================================================

/**
 * Get token list for a chain
 * @param {string} chain - Blockchain
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Token list
 */
async function getTokens(chain, options = {}) {
  const params = new URLSearchParams({
    chain,
    limit: options.limit || 100,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/tokens?${params}`);
  return response.json();
}

/**
 * Get token details
 * @param {string} address - Token address
 * @param {string} chain - Blockchain
 * @returns {Promise<Object>} Token details
 */
async function getToken(address, chain = 'ethereum') {
  const response = await fetch(
    `${BASE_URL}/api/blockchain/token?address=${address}&chain=${chain}`
  );
  return response.json();
}

// =============================================================================
// LAYER 2 & BRIDGES
// =============================================================================

/**
 * Get Layer 2 data
 * @param {Object} options - Query options
 * @returns {Promise<Object>} L2 data
 */
async function getLayer2Data(options = {}) {
  const params = new URLSearchParams({
    network: options.network || 'all',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/l2?${params}`);
  return response.json();
}

/**
 * Get bridge statistics
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Bridge data
 */
async function getBridges(options = {}) {
  const params = new URLSearchParams({
    sourceChain: options.sourceChain || 'all',
    destChain: options.destChain || 'all',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/bridges?${params}`);
  return response.json();
}

// =============================================================================
// NFT DATA
// =============================================================================

/**
 * Get NFT news and trends
 * @param {Object} options - Query options
 * @returns {Promise<Object>} NFT news
 */
async function getNFTNews(options = {}) {
  const params = new URLSearchParams({
    limit: options.limit || 20,
    category: options.category || 'all',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/nft/news?${params}`);
  return response.json();
}

/**
 * Get NFT collection data
 * @param {string} collection - Collection slug or address
 * @returns {Promise<Object>} Collection data
 */
async function getNFTCollection(collection) {
  const response = await fetch(`${BASE_URL}/api/blockchain/nft/collection?id=${collection}`);
  return response.json();
}

// =============================================================================
// AIRDROPS
// =============================================================================

/**
 * Get upcoming airdrops
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Airdrop list
 */
async function getAirdrops(options = {}) {
  const params = new URLSearchParams({
    status: options.status || 'upcoming',
    chain: options.chain || 'all',
    limit: options.limit || 20,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/airdrops?${params}`);
  return response.json();
}

// =============================================================================
// SECURITY
// =============================================================================

/**
 * Get security alerts for the crypto ecosystem
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Security alerts
 */
async function getSecurityAlerts(options = {}) {
  const params = new URLSearchParams({
    severity: options.severity || 'all',
    limit: options.limit || 20,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/security/alerts?${params}`);
  return response.json();
}

/**
 * Get recent hacks and exploits
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Hacks data
 */
async function getHacks(options = {}) {
  const params = new URLSearchParams({
    limit: options.limit || 20,
    year: options.year || new Date().getFullYear(),
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/blockchain/security/hacks?${params}`);
  return response.json();
}

/**
 * Check rug pull risk for a token
 * @param {string} address - Token address
 * @param {string} chain - Blockchain
 * @returns {Promise<Object>} Risk analysis
 */
async function checkRugPullRisk(address, chain = 'ethereum') {
  const response = await fetch(
    `${BASE_URL}/api/blockchain/security/rugcheck?address=${address}&chain=${chain}`
  );
  return response.json();
}

/**
 * Get audit information for a project
 * @param {string} project - Project name or contract address
 * @returns {Promise<Object>} Audit data
 */
async function getAudits(project) {
  const response = await fetch(`${BASE_URL}/api/blockchain/security/audits?project=${project}`);
  return response.json();
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  // On-chain
  getOnChainData,
  getGasPrices,
  getTokenHolders,
  getWhaleMovements,
  
  // DeFi
  getDefiTVL,
  getYields,
  getStaking,
  
  // Tokens
  getTokens,
  getToken,
  
  // L2 & Bridges
  getLayer2Data,
  getBridges,
  
  // NFT
  getNFTNews,
  getNFTCollection,
  
  // Airdrops
  getAirdrops,
  
  // Security
  getSecurityAlerts,
  getHacks,
  checkRugPullRisk,
  getAudits,
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n=== Blockchain & On-Chain API Examples ===\n');
  
  try {
    // Gas prices
    console.log('1. Fetching Ethereum gas prices...');
    const gas = await getGasPrices('ethereum');
    console.log(`   Fast: ${gas.fast || 'N/A'} gwei`);
    
    // DeFi TVL
    console.log('2. Fetching DeFi TVL...');
    const tvl = await getDefiTVL({ chain: 'ethereum' });
    console.log(`   Total TVL: $${tvl.totalTvl?.toLocaleString() || 'N/A'}`);
    
    // Whale movements
    console.log('3. Fetching whale movements...');
    const whales = await getWhaleMovements({ minValue: 10000000, limit: 5 });
    console.log(`   Found ${whales.data?.length || 0} whale transactions`);
    
    // Yields
    console.log('4. Fetching yield farming opportunities...');
    const yields = await getYields({ chain: 'ethereum', limit: 5 });
    console.log(`   Found ${yields.data?.length || 0} yield opportunities`);
    
    // L2 data
    console.log('5. Fetching Layer 2 data...');
    const l2 = await getLayer2Data();
    console.log(`   L2 networks: ${l2.networks?.length || 0}`);
    
    // Security alerts
    console.log('6. Fetching security alerts...');
    const alerts = await getSecurityAlerts({ limit: 5 });
    console.log(`   Found ${alerts.data?.length || 0} security alerts`);
    
    // Recent hacks
    console.log('7. Fetching recent hacks...');
    const hacks = await getHacks({ limit: 5 });
    console.log(`   Found ${hacks.data?.length || 0} recent exploits`);
    
    // Airdrops
    console.log('8. Fetching upcoming airdrops...');
    const airdrops = await getAirdrops({ limit: 5 });
    console.log(`   Found ${airdrops.data?.length || 0} upcoming airdrops`);
    
    console.log('\n✅ All blockchain examples completed!\n');
  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  runExamples();
}

File: market-data/crypto-news/examples/python/blockchain.py
#!/usr/bin/env python3
"""
Blockchain & On-Chain API Examples - Python
Free Crypto News API - https://github.com/nirholas/free-crypto-news

Examples for blockchain and on-chain data endpoints.
"""

import requests
import json
from typing import Optional, List

BASE_URL = "https://cryptocurrency.cv"


# =============================================================================
# GET /api/nft - NFT News
# =============================================================================

def get_nft_news(limit: int = 20, collection: Optional[str] = None) -> dict:
    """
    Get NFT-related news.
    
    Args:
        limit: Number of articles
        collection: Filter by collection
    
    Returns:
        NFT news articles
    """
    params = {"limit": limit}
    if collection:
        params["collection"] = collection
    
    response = requests.get(f"{BASE_URL}/api/nft", params=params)
    return response.json()


# =============================================================================
# GET /api/onchain - On-Chain Data
# =============================================================================

def get_onchain_data(chain: str = "ethereum", metric: str = "transactions") -> dict:
    """
    Get on-chain metrics.
    
    Args:
        chain: Blockchain (ethereum, bitcoin, solana, etc.)
        metric: Metric type (transactions, fees, active_addresses)
    
    Returns:
        On-chain metrics
    """
    params = {"chain": chain, "metric": metric}
    response = requests.get(f"{BASE_URL}/api/onchain", params=params)
    return response.json()


# =============================================================================
# GET /api/onchain/gas - Gas Prices
# =============================================================================

def get_gas_prices(chain: str = "ethereum") -> dict:
    """
    Get current gas prices.
    
    Args:
        chain: Blockchain
    
    Returns:
        Gas price data
    """
    response = requests.get(f"{BASE_URL}/api/onchain/gas", params={"chain": chain})
    return response.json()


# =============================================================================
# GET /api/onchain/holders - Token Holders
# =============================================================================

def get_token_holders(token: str, limit: int = 100) -> dict:
    """
    Get top token holders.
    
    Args:
        token: Token address or symbol
        limit: Number of holders
    
    Returns:
        Top holder list
    """
    params = {"token": token, "limit": limit}
    response = requests.get(f"{BASE_URL}/api/onchain/holders", params=params)
    return response.json()


# =============================================================================
# GET /api/onchain/whales - Whale Movements
# =============================================================================

def get_whale_movements(asset: str = "BTC", min_value: int = 1000000) -> dict:
    """
    Get whale wallet movements.
    
    Args:
        asset: Asset symbol
        min_value: Minimum transaction value
    
    Returns:
        Whale movements
    """
    params = {"asset": asset, "min_value": min_value}
    response = requests.get(f"{BASE_URL}/api/onchain/whales", params=params)
    return response.json()


# =============================================================================
# GET /api/onchain/defi - DeFi TVL
# =============================================================================

def get_defi_tvl(protocol: Optional[str] = None) -> dict:
    """
    Get DeFi Total Value Locked.
    
    Args:
        protocol: Filter by protocol
    
    Returns:
        TVL data
    """
    params = {}
    if protocol:
        params["protocol"] = protocol
    
    response = requests.get(f"{BASE_URL}/api/onchain/defi", params=params)
    return response.json()


# =============================================================================
# GET /api/tokens - Token Data
# =============================================================================

def get_tokens(chain: str = "ethereum", limit: int = 100) -> dict:
    """
    Get token list.
    
    Args:
        chain: Blockchain
        limit: Number of tokens
    
    Returns:
        Token list
    """
    params = {"chain": chain, "limit": limit}
    response = requests.get(f"{BASE_URL}/api/tokens", params=params)
    return response.json()


def get_token(address: str, chain: str = "ethereum") -> dict:
    """
    Get token details.
    
    Args:
        address: Token contract address
        chain: Blockchain
    
    Returns:
        Token details
    """
    response = requests.get(f"{BASE_URL}/api/tokens/{address}", 
                           params={"chain": chain})
    return response.json()


# =============================================================================
# GET /api/staking - Staking Data
# =============================================================================

def get_staking(asset: Optional[str] = None) -> dict:
    """
    Get staking data and yields.
    
    Args:
        asset: Filter by asset
    
    Returns:
        Staking opportunities
    """
    params = {}
    if asset:
        params["asset"] = asset
    
    response = requests.get(f"{BASE_URL}/api/staking", params=params)
    return response.json()


# =============================================================================
# GET /api/layer2 - Layer 2 Data
# =============================================================================

def get_layer2_data() -> dict:
    """
    Get Layer 2 scaling solutions data.
    
    Returns:
        L2 data (Arbitrum, Optimism, etc.)
    """
    response = requests.get(f"{BASE_URL}/api/layer2")
    return response.json()


# =============================================================================
# GET /api/bridges - Cross-Chain Bridges
# =============================================================================

def get_bridges() -> dict:
    """
    Get cross-chain bridge data.
    
    Returns:
        Bridge TVL and volume
    """
    response = requests.get(f"{BASE_URL}/api/bridges")
    return response.json()


# =============================================================================
# GET /api/yields - DeFi Yields
# =============================================================================

def get_yields(chain: Optional[str] = None, 
               min_apy: Optional[float] = None) -> dict:
    """
    Get DeFi yield opportunities.
    
    Args:
        chain: Filter by chain
        min_apy: Minimum APY
    
    Returns:
        Yield opportunities
    """
    params = {}
    if chain:
        params["chain"] = chain
    if min_apy:
        params["min_apy"] = min_apy
    
    response = requests.get(f"{BASE_URL}/api/yields", params=params)
    return response.json()


# =============================================================================
# GET /api/airdrops - Airdrop Opportunities
# =============================================================================

def get_airdrops(status: str = "active") -> dict:
    """
    Get crypto airdrops.
    
    Args:
        status: Filter by status (active, upcoming, ended)
    
    Returns:
        Airdrop opportunities
    """
    response = requests.get(f"{BASE_URL}/api/airdrops", params={"status": status})
    return response.json()


# =============================================================================
# GET /api/security - Security Alerts
# =============================================================================

def get_security_alerts(severity: Optional[str] = None, 
                        limit: int = 20) -> dict:
    """
    Get security alerts (hacks, vulnerabilities).
    
    Args:
        severity: Filter by severity (high, medium, low)
        limit: Number of alerts
    
    Returns:
        Security alerts
    """
    params = {"limit": limit}
    if severity:
        params["severity"] = severity
    
    response = requests.get(f"{BASE_URL}/api/security", params=params)
    return response.json()


# =============================================================================
# GET /api/hacks - Hack Reports
# =============================================================================

def get_hacks(limit: int = 20) -> dict:
    """
    Get crypto hack reports.
    
    Args:
        limit: Number of reports
    
    Returns:
        Hack reports
    """
    response = requests.get(f"{BASE_URL}/api/hacks", params={"limit": limit})
    return response.json()


# =============================================================================
# GET /api/rugpull - Rug Pull Detection
# =============================================================================

def check_rugpull(token: str) -> dict:
    """
    Check token for rug pull indicators.
    
    Args:
        token: Token address
    
    Returns:
        Risk assessment
    """
    response = requests.get(f"{BASE_URL}/api/rugpull", params={"token": token})
    return response.json()


# =============================================================================
# GET /api/audit - Smart Contract Audits
# =============================================================================

def get_audits(protocol: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get smart contract audits.
    
    Args:
        protocol: Filter by protocol
        limit: Number of audits
    
    Returns:
        Audit reports
    """
    params = {"limit": limit}
    if protocol:
        params["protocol"] = protocol
    
    response = requests.get(f"{BASE_URL}/api/audit", params=params)
    return response.json()


# =============================================================================
# COMPLETE EXAMPLES
# =============================================================================

if __name__ == "__main__":
    print("\n" + "="*60)
    print("FREE CRYPTO NEWS API - BLOCKCHAIN & ON-CHAIN EXAMPLES")
    print("="*60)
    
    # 1. NFT News
    print("\n🎨 1. NFT News")
    nft = get_nft_news(limit=5)
    print(f"   NFT News: {nft}")
    
    # 2. On-Chain Metrics
    print("\n⛓️ 2. Ethereum On-Chain Data")
    onchain = get_onchain_data(chain="ethereum", metric="transactions")
    print(f"   On-Chain: {onchain}")
    
    # 3. Gas Prices
    print("\n⛽ 3. ETH Gas Prices")
    gas = get_gas_prices("ethereum")
    print(f"   Gas: {gas}")
    
    # 4. Whale Movements
    print("\n🐋 4. BTC Whale Movements")
    whales = get_whale_movements("BTC", min_value=5000000)
    print(f"   Whales: {whales}")
    
    # 5. DeFi TVL
    print("\n🔒 5. DeFi TVL")
    tvl = get_defi_tvl()
    print(f"   TVL: {tvl}")
    
    # 6. Staking
    print("\n🥩 6. Staking Opportunities")
    staking = get_staking()
    print(f"   Staking: {staking}")
    
    # 7. Layer 2
    print("\n🔷 7. Layer 2 Data")
    l2 = get_layer2_data()
    print(f"   L2: {l2}")
    
    # 8. Yields
    print("\n💰 8. DeFi Yields (>10% APY)")
    yields = get_yields(min_apy=10)
    print(f"   Yields: {yields}")
    
    # 9. Airdrops
    print("\n🪂 9. Active Airdrops")
    airdrops = get_airdrops(status="active")
    print(f"   Airdrops: {airdrops}")
    
    # 10. Security
    print("\n🔐 10. Security Alerts")
    security = get_security_alerts(severity="high", limit=5)
    print(f"   Security: {security}")
    
    # 11. Hacks
    print("\n⚠️ 11. Recent Hacks")
    hacks = get_hacks(limit=5)
    print(f"   Hacks: {hacks}")
    
    # 12. Audits
    print("\n📋 12. Smart Contract Audits")
    audits = get_audits(limit=5)
    print(f"   Audits: {audits}")
    
    print("\n" + "="*60)
    print("All blockchain & on-chain examples completed!")
    print("="*60)
